psl-cli/src/test/java/org/linqs/psl/cli/CLITest.java

```diff
@@ -95,6 +107,156 @@ public void run(String modelPath, String dataPath, String loggingLevel, List<Str
+        File actionFile = new File(actionPath);
+
+        // Set the server command line arguments.
+        serverArgs.add("--" + CommandLineLoader.OPERATION_INFER_LONG);
+        serverArgs.add("SGDOnlineInference");
+
+        serverArgs.add("--" + CommandLineLoader.OPTION_MODEL_LONG);
+        serverArgs.add(modelPath);
+
+        serverArgs.add("--" + CommandLineLoader.OPTION_DATA_LONG);
+        serverArgs.add(dataPath);
+
+        serverArgs.add("--" + CommandLineLoader.OPTION_OUTPUT_DIR_LONG);
+        serverArgs.add(outDir);
+
+        serverArgs.add("-" + CommandLineLoader.OPTION_PROPERTIES);
+        serverArgs.add("log4j.threshold=" + loggingLevel);
+
+        // Start the server.
+        Runnable serverTask = new Runnable() {
+            @Override
+            public void run() {
+                Launcher.main(serverArgs.toArray(new String[0]), true);
+            }
+        };
+        Thread serverThread = new Thread(serverTask);
+        serverThread.start();
+
+        // Wait until the server is ready to accept client connections.
+        waitForServerTempFile();
+
+        // Set the client command line arguments.
+        clientArgs.add("--" + CommandLineLoader.OPERATION_ONLINE_CLIENT_LONG);
+
+        clientArgs.add("--" + CommandLineLoader.OPTION_OUTPUT_DIR_LONG);
+        clientArgs.add(outDir);
+
+        clientArgs.add("-" + CommandLineLoader.OPTION_PROPERTIES);
+        clientArgs.add("log4j.threshold=" + loggingLevel);
+
+        if (additionalArgs != null) {
+            clientArgs.addAll(additionalArgs);
+        }
+
+        // Start the client.
+        ClientTask clientTask = new ClientTask(actionFile, clientArgs);
+        Thread clientThread = new Thread(clientTask);
+        clientThread.start();
+
+        try {
+            clientThread.join();
+        } catch (InterruptedException ex) {
+            throw new RuntimeException(ex);
+        }
+
+        try {
+            serverThread.join();
+        } catch (InterruptedException ex) {
+            throw new RuntimeException(ex);
+        }
+
+        validate();
+
+        return clientTask.getClientOutput();
+    }
+
+    /**
+     * Wait for the online server to drop its temporary file.
+     */
+    public void waitForServerTempFile() {
+        boolean tempFileExists = false;
+
+        while (!tempFileExists) {
+            File tempDir = new File(serverTempFileDir);
+            File[] tempDirFiles = tempDir.listFiles();
+
+            for (File tempDirFile : tempDirFiles) {
+                if (tempDirFile.isFile()) {
+                    String serverTempFileName = tempDirFile.getName();
+                    if (serverTempFileName.startsWith(SERVER_TEMP_FILE_NAME_PREFIX)
+                            && serverTempFileName.endsWith(SERVER_TEMP_FILE_NAME_SUFFIX)) {
+                        tempFileExists = true;
+                        break;
+                    }
+                }
+            }
```

Comment:
This is too complex, just agree on the path of the temp file ahead of time (in static or `@Before`).